src/digest/digest.cr

```diff
@@ -125,21 +125,79 @@ abstract class Digest
 
   # Returns the final digest output.
   #
-  # This method can only be called once and raises `FinalizedError` on subsequent calls.
+  # `final` or `hexfinal` can only be called once and raises `FinalizedError` on subsequent calls.
   #
   # NOTE: `.dup.final` call may be used to get an intermediate hash value.
   def final : Bytes
     dst = Bytes.new digest_size
     final dst
   end
 
+  # Puts the final digest output into `dst`.
+  #
+  # Faster than the `Bytes` allocating version.
+  # Use when hashing in loops.
+  #
+  # `final` or `hexfinal` can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final(dst)` call may be used to get an intermediate hash value.
   def final(dst : Bytes) : Bytes
     check_finished
     @finished = true
     final_impl dst
     dst
   end
 
+  # Returns a hexadecimal-encoded digest in a new `String`.
+  #
+  # `final` or `hexfinal` can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.hexfinal` call may be used to get an intermediate hash value.
+  def hexfinal : String
+    dsize = digest_size
+    string_size = dsize * 2
+    String.new(string_size) do |buffer|
+      tmp = Slice.new(buffer + dsize, dsize)
+      final tmp
+      tmp.hexstring buffer
+      {string_size, string_size}
+    end
+  end
+
+  # Writes a hexadecimal-encoded digest to `dst`.
+  #
+  # Faster than the `String` allocating version.
+  # Use when hashing in loops.
+  #
+  # `final` or `hexfinal` can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final` call may be used to get an intermediate hash value.
+  def hexfinal(dst : Bytes) : Nil
+    dsize = digest_size
+    unless dst.bytesize == dsize * 2
+      raise ArgumentError.new("Incorrect dst size: #{dst.bytesize}, expected: #{dsize * 2}")
+    end
+
+    tmp = dst[dsize, dsize]
+    final tmp
+    tmp.hexstring dst
+  end
+
+  # Writes a hexadecimal-encoded digest to `IO`.
+  #
+  # Faster than the `String` allocating version.
+  #
+  # `final` or `hexfinal` can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final` call may be used to get an intermediate hash value.
+  def hexfinal(io : IO) : Nil
+    sary = uninitialized StaticArray(UInt8, 128)
+    tmp = sary.to_slice[0, digest_size * 2]
```

Comment:
We should address the size restriction in this method. It's not possible to use the same optimization as in the `String` overload.
I think it's probably best to raise a meaningful error if `digest_size * 2 > 128` and mention this in the documentations (that implementing types can override this method for a larger buffer size).